bindings/rust/bench/src/harness.rs

```diff
@@ -1,163 +1,356 @@
+    }
+}
+
+pub fn get_cert_path(pem_type: PemType, sig_type: SigType) -> String {
+    format!(
+        "certs/{}/{}",
+        sig_type.get_dir_name(),
+        pem_type.get_filename()
+    )
+}
 
 pub fn read_to_bytes(pem_type: PemType, sig_type: SigType) -> Vec<u8> {
     read_to_string(get_cert_path(pem_type, sig_type))
         .unwrap()
         .into_bytes()
 }
 
-#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+#[derive(Clone, Copy)]
 pub enum Mode {
     Client,
     Server,
 }
 
-#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
+#[derive(Clone, Copy, Default, EnumIter, Eq, PartialEq)]
 pub enum HandshakeType {
     #[default]
     ServerAuth,
     MutualAuth,
 }
 
+impl Debug for HandshakeType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", match self {
+            Self::ServerAuth => "",
+            Self::MutualAuth => "mTLS",
+        })
+    }
+}
+
 // these parameters were the only ones readily usable for all three libaries:
 // s2n-tls, rustls, and openssl
 #[allow(non_camel_case_types)]
-#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
+#[derive(Clone, Copy, Debug, Default, EnumIter, Eq, PartialEq)]
 pub enum CipherSuite {
     #[default]
     AES_128_GCM_SHA256,
     AES_256_GCM_SHA384,
 }
 
-#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
-pub enum ECGroup {
-    SECP256R1,
+#[derive(Clone, Copy, Default, EnumIter)]
+pub enum KXGroup {
+    Secp256R1,
     #[default]
     X25519,
 }
 
-#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
-pub enum SigType {
-    Rsa2048,
-    Rsa3072,
-    Rsa4096,
-    #[default]
-    Ec384,
-}
-
-impl SigType {
-    pub fn get_dir_name(&self) -> &str {
-        match self {
-            SigType::Rsa2048 => "rsa2048",
-            SigType::Rsa3072 => "rsa3072",
-            SigType::Rsa4096 => "rsa4096",
-            SigType::Ec384 => "ec384",
-        }
+impl Debug for KXGroup {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", match self {
+            Self::Secp256R1 => "secp256r1",
+            Self::X25519 => "x25519",
+        })
     }
 }
 
-#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
+#[derive(Clone, Copy, Debug, Default)]
 pub struct CryptoConfig {
     pub cipher_suite: CipherSuite,
-    pub ec_group: ECGroup,
+    pub kx_group: KXGroup,
     pub sig_type: SigType,
 }
 
 impl CryptoConfig {
-    pub fn new(cipher_suite: CipherSuite, ec_group: ECGroup, sig_type: SigType) -> Self {
+    pub fn new(cipher_suite: CipherSuite, kx_group: KXGroup, sig_type: SigType) -> Self {
         Self {
             cipher_suite,
-            ec_group,
+            kx_group,
             sig_type,
         }
     }
 }
 
-pub trait TlsBenchHarness: Sized {
-    /// Default harness
-    fn default() -> Result<Self, Box<dyn Error>> {
-        Self::new(CryptoConfig::default(), HandshakeType::default(), ConnectedBuffer::default())
+pub trait TlsConnection: Sized {
```

Comment:
?

Reply:
Oh oops, sorry for missing this first time around! I'd still prefer the file to be called `harness.rs`, because this file contains a lot of different stuff that isn't just `TlsConnection` or `TlsConnPair`; it contains everything required to make that trait and struct work, such as the config options and `ConnectedBuffer`.